### DISCUSSION TITLE
Reboot with fallback

### DISCUSSION THREAD

**Discussion by Ginodel** [JAndrassy/ArduinoOTA#73] [Ideas]
Nice work Jandrassy... Do you know how to download new sketch, flash and reboot, but when execution fails (for any reason to come back to the old version ? 

This should be very nice and probably somebody worked this out ... Any idea ?
I am talking here about implementation for SAMD21 processor (MKR zero board) eventually with additional flash memory.

**Comment by JAndrassy** [comment]
the MCU binaries are linked for exact addresses. some MCU CPU (esp for example) can set the relative address, then the binary can be twice in flash and run from any copy.
if the binary is successful stored in upper half of the flash, there is a very little chance of errors at copying to run location.

**Comment by Ginodel** [comment]
Thanks for your reply Jandrassy,

As i understood from your library you  can upload new firmware to external flash.  Is there no way in the bootloader to tell him to start execution of code from that external flash instead of the internal flash ? 

If in that procedure the flag telling to boot from external flash is cancelled, and execution is failing, then with the watchdog you will come back to the code in internal flash.

For sure that should need an adapted bootloader.  i know ..

Maybe i'm telling stupid things because i have not enough knowledge to tell if this is possible with an adapted bootloader.

Another possibility is like you said… using a different MCU … I was looking yesterday into the ESP32 and indeed in the ESP32 IDF sdk is doing all this ….. for my project i do not need WIFI, BLE … and honnestly i am a bit affraid to switch from SAMD to ESP32 although i see more things that are very nice … as FreeRTOS for multitasking, double core, a lot more memory (flash and Ram), secure boot …. but i may not run into other problems because of the switch as i use Modbus TCP master and slave, CAN interface, etc etc 

I do not have any experience with ESP32 at this moment so difficult for me to tell if the switch should be straight forward. On top I have very little time to experiment because project has to be worked out in a small period of time.

Any suggestions ?  Experience with ESP32 ?  Your input should be appreciated ….Thanks

**Reply by JAndrassy** [reply]
can SAMD21 run from external flash? the bootloader would have to do the switching, not this library

**Comment by Ginodel** [comment]
Hi Jandrassy,

That's what i told … it would need an adapted bootloader and has indeed nothing to do with this library. 
Any reaction on the second part of my reply … using ESP32 .??? 

Thanks for response

**Reply by JAndrassy** [reply]
I have very little experience with esp32. I think the ESP32 Arduino doesn't use the SDK functions for OTA, but does a similar write-to-store then copy approach as this library.
